accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+
+Usage is generally transparent to the user. When creating a new process, the current process architectural affinity is used to select the best architecture to use for the new process.
+
+The user also has explicit architecture control through the use of the `arch` utility. The user can select a specific architecture when executing a command from the command line i.e. `arch -arch x86_64 <command>`.
+
+While Rosetta 2 is supported for running `x64` code on Apple Silicon, the reverse emulator is not currently shipping.
+
+#### Universal Binaries
+
+For macOS, universal are the default approach. Most binaries are multi-arch. For single arch binaries the kernel can detect and automatically switch the process to run in the correct architecture, either native or Rosetta 2 emulation of x64.
+
+A universal binary can be created simply by combining two or more architecture specific binaries into a universal binary using the `lipo` tool. The process is reversible too. A universal binary can be `thinned` to extract a single arch binaries.
+
+Given both architectures, universal binaries are relatively easy to create. The general syntax would be `lipo <binaryArchA> <binaryArchB> -create -output <binaryUniversal>`
+
+#### Apple Store requires universal binaries
+
+I have heard that publishing native Apple Silicon (or Apps supporting multiple architectures) apps to the Apple store will require them to be published as Universal binaries. I have not been able to find public documentation of this, so the details are not clear.
+
+This document is assumes this requirement is reasonably accurate.
+
+### Security Enhancements
+
+#### Write No Execute
+
+Another major change with the Apple Silicon is the transition to a new prohibition of pages simultaneously marker writable and executable. A new API was introduced for JIT usage to allow a single thread to switch a page from executable to writable or vice-versa.
+
+This requirement currently only applies when executing as Apple Silicon native code.
+
+This is not expected to impact our customers as it needs to be handled automatically by our run-times.
+
+#### Tightened Signing Requirements
+
+When running on Apple Silicon unsigned binaries will get killed automatically by the kernel. A simple non-obvious message will be printed on the console (`zsh killed: helloWorld`).
+
+To allow a native shared library to be loaded, it will also need to be signed.
+
+When running the equivalent unsigned binary from a Rosetta 2 emulated process the process is not killed. When running a x64 process on Apple Silicon the process was not killed either.
+
+XCode 12 adhoc signs native binaries. These anonymous signature are sufficient to prevent these from being killed (on the machine that compiled them at least.)
+
+I haven't found the documentation exactly how and what changed. These behavior descriptions have been through experimentation and debugging.
```

Comment:
Have you seen the work we have done on this? https://docs.microsoft.com/en-us/dotnet/core/install/macos-notarization-issues

> When running the equivalent unsigned binary from a Rosetta 2 emulated process the process is not killed.

I think that this is temporary to allow easier transition and it will be locked down.

Reply:
@odhanson Might be able to provide more context.

Reply:
I will try gather more information both about the signing and the Apple store requirements and comment back here.

Reply:
Here is a link:
https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-universal-apps-release-notes

Code Signing
New Features
New in macOS 11 on Macs with Apple silicon, and starting in macOS Big Sur 11 beta 6, the operating system enforces that any executable must be signed before it’s allowed to run. There isn’t a specific identity requirement for this signature: a simple ad-hoc signature is sufficient. This new behavior doesn’t change the long-established policy that our users and developers can run arbitrary code on their Macs, and is designed to simplify the execution policies on Macs with Apple silicon and enable the system to better detect code modifications. This new policy doesn’t apply to translated x86 binaries running under Rosetta 2, nor does it apply to macOS 11 running on Intel-based platforms.